lib/date_time/date_time_core.c

```diff
@@ -0,0 +1,268 @@
+
+	/* Update time source to the event if not requesting previous time source */
+	if (time_source != DATE_TIME_EVT_TYPE_PREVIOUS) {
+		evt.type = time_source;
+	}
+
+	if (app_evt_handler != NULL) {
+		app_evt_handler(&evt);
+	}
+}
+
+static void date_time_core_schedule_update(void)
+{
+	/* If periodic updates are requested in the configuration
+	 * and next update hasn't been already scheduled.
+	 */
+	if (CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS > 0 &&
+	    !k_work_delayable_is_pending(&time_work)) {
+
+		LOG_DBG("New date time update in: %d seconds",
+			CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS);
+
+		k_work_schedule(&time_work, K_SECONDS(CONFIG_DATE_TIME_UPDATE_INTERVAL_SECONDS));
+	}
+}
+
+static void date_time_update_thread(void)
+{
+	int err;
+
+	while (true) {
+		k_sem_take(&time_fetch_sem, K_FOREVER);
+
+		LOG_DBG("Updating date time UTC...");
+
+		err = date_time_core_current_check();
+		if (err == 0) {
+			LOG_DBG("Using previously obtained time");
+			date_time_core_schedule_update();
+			date_time_core_notify_event(DATE_TIME_EVT_TYPE_PREVIOUS);
+			continue;
+		}
+
+#if defined(CONFIG_DATE_TIME_MODEM)
+		LOG_DBG("Getting time from cellular network");
+		int64_t date_time_ms_modem = 0;
+
+		err = date_time_modem_get(&date_time_ms_modem);
+		if (err == 0) {
+			date_time_core_store(date_time_ms_modem, DATE_TIME_OBTAINED_MODEM);
+			continue;
+		}
+#endif
+#if defined(CONFIG_DATE_TIME_NTP)
+		LOG_DBG("Getting time from NTP server");
+		int64_t date_time_ms_ntp = 0;
+
+		err = date_time_ntp_get(&date_time_ms_ntp);
+		if (err == 0) {
+			date_time_core_store(date_time_ms_ntp, DATE_TIME_OBTAINED_NTP);
+			continue;
+		}
+#endif
+		LOG_DBG("Did not get time from any time source");
+
+		date_time_core_notify_event(DATE_TIME_NOT_OBTAINED);
+	}
+}
+
+K_THREAD_DEFINE(time_thread, CONFIG_DATE_TIME_THREAD_STACK_SIZE,
+		date_time_update_thread, NULL, NULL, NULL,
+		K_LOWEST_APPLICATION_THREAD_PRIO, 0, 0);
+
+static void date_time_handler(struct k_work *work)
+{
+	k_sem_give(&time_fetch_sem);
+}
+
+void date_time_lte_ind_handler(const struct lte_lc_evt *const evt)
+{
+#if defined(CONFIG_DATE_TIME_AUTO_UPDATE) && defined(CONFIG_LTE_LINK_CONTROL)
+	switch (evt->type) {
+	case LTE_LC_EVT_NW_REG_STATUS:
+
+		switch (evt->nw_reg_status) {
+		case LTE_LC_NW_REG_REGISTERED_EMERGENCY:
+		case LTE_LC_NW_REG_REGISTERED_HOME:
+		case LTE_LC_NW_REG_REGISTERED_ROAMING:
+			if (!date_time_is_valid() && !k_work_delayable_is_pending(&time_work)) {
+				k_work_schedule(&time_work, K_SECONDS(1));
+			}
+			break;
+#if defined(CONFIG_DATE_TIME_MODEM)
+		case LTE_LC_NW_REG_SEARCHING:
+			date_time_modem_xtime_subscribe();
+			break;
+#endif
+		default:
+			break;
+		}
+		break;
+	default:
+		break;
+	}
+#endif /* defined(CONFIG_DATE_TIME_AUTO_UPDATE) && defined(CONFIG_LTE_LINK_CONTROL) */
+}
+
+void date_time_core_init(void)
+{
+	if (IS_ENABLED(CONFIG_DATE_TIME_MODEM) && IS_ENABLED(CONFIG_DATE_TIME_AUTO_UPDATE) &&
+	    IS_ENABLED(CONFIG_AT_MONITOR_SYS_INIT)) {
```

Comment:
This seems to be missing a NOT?
```suggestion
	    !IS_ENABLED(CONFIG_AT_MONITOR_SYS_INIT)) {
```

Reply:
Thanks, fixed.